src/com/palmergames/bukkit/towny/utils/AreaSelectionUtil.java

```diff
@@ -72,101 +86,178 @@
+	 * @return List&lt;WorldCoord&gt; of {@link com.palmergames.bukkit.towny.object.WorldCoord}.
+	 * @throws TownyException - Thrown when invalid radii are given.
+	 */
+	private static List<WorldCoord> selectWorldCoordAreaRect(int available, WorldCoord pos, String[] args) throws TownyException {
 
 		List<WorldCoord> out = new ArrayList<>();
-		if (pos.getTownyWorld().isClaimable()) {
-			if (args.length > 0) {
-				int r = 0, available = 1000;
-
-				if (owner instanceof Town) {
-					Town town = (Town) owner;
-					available = TownySettings.getMaxTownBlocks(town) - town.getTownBlocks().size();
-				} else if (owner instanceof Resident) {
-					available = TownySettings.getMaxResidentPlots((Resident) owner);
-				}
-				
+		if (args.length > 0) {
+			int r = MAX_RECT_RADIUS;  // The radius of the claim.
+
+			/*
+			 *  Area selections are capped at a 15 radius which should be a 31x31 (or a square with a side of 496 blocks in length.)  
+			 *  Players need a permission node to use area claims and the max radius usable defaults to 4 (set in the config.)
+			 */
+
+			if (args[0].equalsIgnoreCase("auto")) {
 				
+				/*
+				 * Select everything possible in a rectangle shape.
+				 */
 
-				if (args[0].equalsIgnoreCase("auto")) {
-					// Attempt to select outwards until no town blocks remain
+				if (TownySettings.getMaxClaimRadiusValue() > 0) 
+					r = Math.min(r, TownySettings.getMaxClaimRadiusValue());
 
-					while (available - Math.pow((r + 1) * 2 - 1, 2) >= 0)
-						r += 1;
+			} else {
+			
+				/*
+				 * Select an area that will claim a perfect square shape, using a given radius
+				 * or the reduced radius that will give a perfect square.
+				 */
+				
+				try {
+					r = Integer.parseInt(args[0]);
+				} catch (NumberFormatException e) {
+					throw new TownyException(Translation.of("msg_err_invalid_radius"));
+				}
+				if (TownySettings.getMaxClaimRadiusValue() > 0 && r > TownySettings.getMaxClaimRadiusValue())
+					throw new TownyException(Translation.of("msg_err_invalid_radius_number", TownySettings.getMaxClaimRadiusValue()));
 
-				} else {
-					try {
-						r = Integer.parseInt(args[0]);
-					} catch (NumberFormatException e) {
-						throw new TownyException(Translation.of("msg_err_invalid_radius"));
+				// Calculate how many TownBlocks are needed to claim the radius. 
+				// Start blocks at 0 if the plot is unclaimed.
+				int neededBlocks = pos.getTownBlock().hasTown() ? 0 : 1;
+				neededBlocks += 4 * r * (r + 1);
+				
+				// Rethink how much of a radius will be used, as there's not enough available TownBlocks.
+				if (neededBlocks > available) {
+					r = 1; 
+					int total = 0;
+					while (available - ((r * 8) + total) >= 0) { // Radius 1 grabs 8 blocks, 2 grabs 16 more requiring 24, 3 grabs 24 more requiring 48, etc...
+						total += r * 8;
+						r++;
 					}
+					r--; // Finally reduce the radius by 1 (so that we have a perfect ring of claims,) and replace the original r.
 				}
```

Comment:
Looking at it again, all of this can be probably be done in one go.
```java
int needed = pos.getTownBlock().hasTown() ? 0 : 1;
int claimRadius = 1;
while (claimRadius <= r && needed < available) {
    needed += (claimRadius * 8)
    claimRadius++;
}
// Claim Radius will always overshoot by 1
r = claimRadius - 1;
```